layerindex/update.py

```diff
@@ -249,57 +265,77 @@ def main():
         if not lockfile:
             logger.error("Layer index lock timeout expired")
             sys.exit(1)
         try:
-            bitbakepath = os.path.join(fetchdir, 'bitbake')
+            # Make sure oe-core is fetched since recipe parsing requires it
+            layerquery_core = LayerItem.objects.filter(comparison=False).filter(name=settings.CORE_LAYER_NAME)
+            if layerquery_core in layerquery:
+                layerquery_fetch = list(layerquery)
+            else:
+                layerquery_fetch = list(layerquery) + list(layerquery_core)
+            # Fetch latest metadata from repositories
+            for layer in layerquery_fetch:
+                # Handle multiple layers in a single repo
+                urldir = layer.get_fetch_dir()
+                repodir = os.path.join(fetchdir, urldir)
+                if layer.vcs_url not in allrepos:
+                    allrepos[layer.vcs_url] = (repodir, urldir, fetchdir, layer.name)
+            # Add bitbake
+            if settings.BITBAKE_REPO_URL not in allrepos:
+                bitbakeitem = LayerItem()
+                bitbakeitem.vcs_url = settings.BITBAKE_REPO_URL
+                bitbakeurldir = bitbakeitem.get_fetch_dir()
+                bitbakepath = os.path.join(fetchdir, bitbakeurldir)
+                allrepos[settings.BITBAKE_REPO_URL] = (bitbakepath, bitbakeurldir, fetchdir, "bitbake")
+
+            (bitbakepath, _, _, _) = allrepos[settings.BITBAKE_REPO_URL]
+            if getattr(settings, 'BITBAKE_PATH', ''):
+                bitbakepath = os.path.join(bitbakepath, settings.BITBAKE_PATH)
 
             if not options.nofetch:
-                # Fetch latest metadata from repositories
-                for layer in layerquery:
-                    # Handle multiple layers in a single repo
-                    urldir = layer.get_fetch_dir()
-                    repodir = os.path.join(fetchdir, urldir)
-                    if not (layer.vcs_url in fetchedrepos or layer.vcs_url in failedrepos):
-                        logger.info("Fetching remote repository %s" % layer.vcs_url)
-                        out = None
-                        try:
-                            if not os.path.exists(repodir):
-                                out = utils.runcmd("git clone %s %s" % (layer.vcs_url, urldir), fetchdir, logger=logger, printerr=False)
-                            else:
-                                out = utils.runcmd("git fetch -p", repodir, logger=logger, printerr=False)
-                        except subprocess.CalledProcessError as e:
-                            logger.error("Fetch of layer %s failed: %s" % (layer.name, e.output))
-                            failedrepos[layer.vcs_url] = e.output
-                            continue
-                        fetchedrepos.append(layer.vcs_url)
+                # Parallel fetching
+                pool = multiprocessing.Pool(int(settings.PARALLEL_JOBS))
+                for url in allrepos:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)